openwisp_notifications/notification_types.py

```diff
@@ -0,0 +1,78 @@
+from django.core.exceptions import ImproperlyConfigured
+from openwisp_notifications.settings import ADDITIONAL_NOTIFICATION_TYPES
+
+DEFAULT_NOTIFICATION_TYPES = {
+    'default': {
+        'level': 'info',
+        'verb': 'default verb',
+        'name': 'Default Type',
+        'email_subject': '[openwisp-notificaton] Default Notification Subject',
+    },
+}
+
+
+def deep_merge_dicts(dict1, dict2):
+    result = dict1.copy()
+    for key, value in dict2.items():
+        if isinstance(value, dict):
+            node = result.get(key, {})
+            result[key] = deep_merge_dicts(node, value)
+        else:
+            result[key] = value
+    return result
+
+
+def get_notification_types():
+    notification_types = deep_merge_dicts(
+        DEFAULT_NOTIFICATION_TYPES, ADDITIONAL_NOTIFICATION_TYPES
+    )
+    validate_notification_type(notification_types)
+    return notification_types
+
+
+def get_notification_choices():
+    notification_types = NOTIFICATION_TYPES
+    choices = []
+    for key in sorted(notification_types.keys()):
+        name = notification_types[key].get('name', key)
+        choices.append((key, name))
+    return choices
+
+
+def validate_notification_type(notification_type):
+    for key, options in notification_type.items():
+        assert 'level' in options
+        assert 'verb' in options
+
+
+def register_notification_type(notification_type):
+    validate_notification_type(notification_type)
+    NOTIFICATION_TYPES.update(notification_type)
+    register_notification_choice(notification_type)
+
+
+def unregister_notification_type(notification_type):
```

Comment:
please document the usage of this function.

Test it in openwisp-monitoring to unregister a notification type that you register just before and ensure the notification type does not appear in the available choices (this test makes sense only after you get `register_notification_type` to  work as expected.

Can notification types be unregistered with something like this right?

```python
unregister_notification_type( 'monitoring.device.up')
```

Reply:
A similar test exists in `openwisp-notifications`, we can later check this in a sample app. Adding this test again will create redundancy that too in different repositories. Please have a look on [this test case](https://github.com/openwisp/openwisp-notifications/pull/19/files#diff-f5244de33ac705050d99b0ff4150c6a8R184-R202)

I will document usage of these functions.

> Can notification types be unregistered with something like this right?
Yes, they can be unregistered like that.

Reply:
> Please have a look on this test case

Are you sure that link is working as intended? It looks like it's not to me.

Reply:
GitHub doesn't open the file if you have already marked it as viewed. Please see those tests here : 

https://github.com/openwisp/openwisp-notifications/blob/927f4fea62e15d2b85ad1485faaadabac03e3248/openwisp_notifications/tests/test_notifications.py#L184-L283